File: RestoreAlgorithm/ShortestPath/PriQueDijkstra.js
const N = 6;
const M = 11;
const Start = 1;

class PriorityQueue {
  constructor() {
    this.que = [];
  }

  enque(node, distance) {
    this.que.push([node, distance]);
  }

  deque() {
    this.que.sort((a, b) => a[1] - b[1]);
    return this.que.splice(0, 1);
  }
}

const distance = Array.from({ length: N + 1 }, () => Infinity);
const graph = Array.from(Array(N + 1), () => []);
const pq = new PriorityQueue();

graph[1].push([2, 2]);
graph[1].push([3, 5]);
graph[1].push([4, 1]);
graph[2].push([3, 3]);
graph[2].push([4, 2]);
graph[3].push([2, 3]);
graph[3].push([6, 5]);
graph[4].push([3, 3]);
graph[4].push([5, 1]);
graph[5].push([3, 1]);
graph[5].push([6, 2]);

distance[Start] = 0;
pq.enque(1, 0);

while (pq.que.length > 0) {
  const [[node, dist]] = pq.deque();

  if (distance[node] < dist) continue;

  for (let [next, nextDis] of graph[node]) {
    if (distance[next] > distance[node] + nextDis) {
      distance[next] = distance[node] + nextDis;
      pq.enque(next, distance[next]);
    }
  }
}

for (let d of distance) {
  if (d !== Infinity) console.log(d);
}

File: RestoreAlgorithm/ShortestPath/PriQueDijkstra.cpp
// ProblemNumber || ProblemName : 우선순위 다익스트라 알고리즘₩₩

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <utility>
#include <queue>

using namespace std;

int main()
{
  ios_base ::sync_with_stdio(false);
  cin.tie(NULL);

  int n, m, s;

  cin >> n >> m;
  cin >> s;

  vector<int> distance(n + 1, 99999);
  vector<vector<pair<int, int>>> graph(n + 1);
  priority_queue<pair<int, int>> que;

  for (int i = 0; i < m; i++)
  {
    int start, end, cost;

    cin >> start >> end >> cost;

    graph[start].push_back({end, cost});
  }

  que.push({0, s});
  distance[s] = 0;

  while (!que.empty())
  {
    int dis = -que.top().first;
    int now = que.top().second;

    que.pop();

    if (distance[now] < dis)
      continue;

    for (int i = 0; i < graph[now].size(); i++)
    {
      int cost = dis + graph[now][i].second;

      if (cost < distance[graph[now][i].first])
      {
        distance[graph[now][i].first] = cost;
        que.push(make_pair(-cost, graph[now][i].first));
      }
    }
  }

  for (int i = 1; i < distance.size(); i++)
  {
    cout << distance[i] << endl;
  }
}
